rllib/utils/actors.py

```diff
@@ -38,27 +39,26 @@ def completed_prefetch(self, blocking_wait=False, max_yield=999):
         for worker, obj_id in self.completed(blocking_wait=blocking_wait):
             self._fetching.append((worker, obj_id))
 
-        remaining = []
-        num_yielded = 0
-        for worker, obj_id in self._fetching:
-            if num_yielded < max_yield:
-                yield (worker, obj_id)
-                num_yielded += 1
-            else:
-                remaining.append((worker, obj_id))
-        self._fetching = remaining
+        for _ in range(max_yield):
+            if not self._fetching:
+                break
+
+            yield self._fetching.popleft()
 
     def reset_workers(self, workers):
         """Notify that some workers may be removed."""
         for obj_id, ev in self._tasks.copy().items():
             if ev not in workers:
                 del self._tasks[obj_id]
                 del self._objects[obj_id]
-        ok = []
-        for ev, obj_id in self._fetching:
+
+        # We want to keep the same deque reference so that we don't suffer from
+        # stale references in generators that are still in flight
+        for _ in range(len(self._fetching)):
```

Comment:
Is it possible for multiple threads to be accessing the TaskPool? Should I add locks to make this operation atomic in case another thread is attempting to prefetch completed tasks while we're resetting the workers?

Reply:
This looks fine -- it should be single threaded.

Reply:
Thanks for confirming